File: linked-lists/linked_list.c
#include "linked_list.h"
#include <stdlib.h>
#include <string.h>

void create_list(List *list, void (*destroy)(void *data))
{
    list->size = 0;
    list->destroy = destroy;
    list->head = NULL;
    list->tail = NULL;
}

void destroy_list(List *list)
{
    void *data;
    while(list_size(list) > 0)
        if(remove_next(list, NULL, (void**)data) == 0 && list->destroy != NULL)
            list->destroy(data);
    memset(list, 0, sizeof(List));
}

int insert_next(List *list, node *elem, const void *data)
{
    node *new_node;
    if((new_node = malloc(sizeof(List))) == NULL)
        return 0; //unsuccessful insertion
    new_node->data = (void*)data;
    if(elem == NULL){
        //insert at beginning
        if(list_size(list) == 0) 
            list->tail = new_node; //make new_node tail
        new_node->next = list->head; //point nodes next to current head
        list->head = new_node;//new_node is new head
    } else {
        //insert after head
        if(elem->next == NULL) //check if elem is tail
            list->tail = new_node; //append at tail of list
        new_node->next = elem->next;
        elem->next = new_node;
    }
    list->size++;
    //successful insertion
    return 1;
}

int remove_next(List *list, node *elem, void **data)
{
    node *old_node;
    //check if list is empty
    if(list_size(list) == 0)
        return 0;
    if(elem == NULL){
        //removing head
        *data = list->head->data; //store heads data
        old_node = list->head; //node to delete set to head
        list->head = list->head->next;//make head whatever old head is pointing to

        if(list_size(list) == 1) //if only node left, make tail
            list->tail = NULL;
    } else {
        //remove elsewhere in list
        if(elem->next == NULL)//if elem is tail, exit
            return 0;
        *data = elem->next->data; //store data of node to be removed
        old_node = elem->next; 
        elem->next = elem->next->next; //point elem->next to oldnode->next

        if(elem->next == NULL)
            list->tail = elem;
    }
    free(old_node);
    list->size--;
    return 1;
}

int list_size(const List *list)
{
    return list->size;
}

int is_head(List *list, const node *elem)
{
    return elem == list->head? 1: 0;
}

int is_tail(const node *elem)
{
    return elem->next == NULL? 1:0;
}

void print_list(List *list, const node *elem)
{
    while(elem != NULL) {
        printf("%p\n", list_data(elem));
        elem = elem->next;
    }
}

int main()
{
    List *list;
    node *elem;

    create_list(list, NULL);
    elem = list_head(list);
    insert_next(list, elem, (int*)5);
    insert_next(list, elem, (int*)3);
    print_list(list, list_head(list));

    elem = list_head(list);
    remove_next(list, elem, &list_data(elem));
    printf("\nRemoving head: \n");
    print_list(list, list_head(list));

    printf("New list: \n");
    insert_next(list, elem, (int*)7);
    insert_next(list, elem, (int*)2);
    print_list(list, list_head(list));

    printf("\nList head: %p\n", list_head(list)->data);
    printf("List tail: %p\n", list_tail(list)->data);
}

File: linked-lists/linked_list.py
#singly linked lists

class Node(object):
    
    def __init__(self, value):

        self._value = value
        self._next_node = None


class SinglyLinkedList(object):

    def __init__(self):
        
        self._head = None
        self._tail = None
        self._size = 0
    
    def insert_node_at_head(self, value):
        
        new_node = Node(value)
        #if list is empty, make new_node head and tail
        if self._head == None:
            self._head = new_node
            self._tail = new_node
            new_node._next_node = None
        else:
            #point new_node to old head and make new_node head
            old_head = self._head
            self._head = new_node
            new_node._next_node = old_head
        self._size += 1

    
    def insert_node_at_tail(self, value):
        
        #make old_tail point to new_node
        new_node = Node(value)
        old_tail = self._tail
        self._tail = new_node
        old_tail._next_node = new_node
        #point new_node to None
        new_node._next_node = None
        self._size += 1

    def remove_node_at_head(self):
        
        if self._head == None:
            print('cannot remove node from an empty list')
            return None
        #point head to None and make next_node the new head
        new_head = self._head._next_node
        self._head._next_node = None
        self._head = new_head
        self._size -= 1

    def remove_node_at_tail(self):
        
        #traverse list to find tail
        if self._head == None:
            print('cannot remove node from an empty list')
            return None
        #start at head, traverse list until find 2nd to last node
        node = self._head
        while node._next_node._next_node != None:
            node = node._next_node
        #point 2nd to last node's next_node to None, making it the tail
        node._next_node = None
        self._tail = node
        self._size -= 1


#    def get_size(self):
#        
#        size = 0
#        node = self._head
#        while node != None:
#            size += 1
#            node = node._next_node
#        return size


    def print_list(self):

        node = self._head
        while node != None:
            print(node._value)
            node = node._next_node

    
    def search_list(self, target):

        node = self._head
        while node != None and node._value != target:
            if node._value == target:
                break
            node = node._next_node
        if node == self._head and node._value != target:
            return None
        return node


llist = SinglyLinkedList()
llist.insert_node_at_head(1)
llist.insert_node_at_head(2)
llist.insert_node_at_head(5)
llist.insert_node_at_tail(7)
llist.insert_node_at_tail(55)
llist.print_list()
print('removing tail')
llist.remove_node_at_tail()
llist.print_list()
print('size of list: ', llist._size)
print('removing head of list')
llist.remove_node_at_head()
llist.print_list()

#search list
find = 1
found_node = llist.search_list(find)
if found_node != None:
    print('Value: ', found_node._value)
    if found_node._next_node != None:
        print('Next node: ', found_node._next_node._value)
    else:
        print(f'node with value {find} is the tail, there is no next node')
else:
    print('Value not found in linked list')
